main/medium/src/reaper.rs

```diff
@@ -5573,6 +5748,461 @@ impl<UsageScope> Reaper<UsageScope> {
+    where
+        UsageScope: MainThreadOnly,
+    {
+        self.require_main_thread();
+        let ptr = self
+            .low
+            .CreateNewMIDIItemInProj(track.as_ptr(), starttime, endtime, &time_in_qn);
+        NonNull::new(ptr).ok_or_else(|| {
+            ReaperFunctionError::new("couldn't create MediaItem (maybe track is invalid)")
+        })
+    }
+
+    /// returns false if there are no plugins on the track that support MIDI programs,
+    /// or if all programs have been enumerated
+    pub fn enum_track_midi_program_names(
+        &self,
+        track: i32,
+        program_number: i32,
+        buffer_size: u32,
+    ) -> Option<ReaperString>
+    where
+        UsageScope: MainThreadOnly,
+    {
+        unsafe {
+            let (program_name, status) = with_string_buffer(buffer_size, |buffer, max_size| {
+                self.low
+                    .EnumTrackMIDIProgramNames(track, program_number, buffer, max_size)
+            });
+            match status {
+                true => Some(program_name),
+                false => None,
+            }
+        }
+    }
+
+    /// returns false if there are no plugins on the track that support MIDI programs,
+    /// or if all programs have been enumerated
+    ///
+    /// # Safety
+    ///
+    /// REAPER can crash if you pass an invalid project or track.
+    pub unsafe fn enum_track_midi_program_names_ex(
+        &self,
+        project: ProjectContext,
+        track: MediaTrack,
+        program_number: i32,
+        buffer_size: u32,
+    ) -> Option<ReaperString>
+    where
+        UsageScope: MainThreadOnly,
+    {
+        self.require_valid_project(project);
+        self.enum_track_midi_program_names_ex_unchecked(project, track, program_number, buffer_size)
+    }
+
+    /// returns false if there are no plugins on the track that support MIDI programs,
+    /// or if all programs have been enumerated
+    ///
+    /// # Safety
+    ///
+    /// REAPER can crash if you pass an invalid project or track.
+    pub unsafe fn enum_track_midi_program_names_ex_unchecked(
+        &self,
+        project: ProjectContext,
+        track: MediaTrack,
+        program_number: i32,
+        buffer_size: u32,
+    ) -> Option<ReaperString>
+    where
+        UsageScope: MainThreadOnly,
+    {
+        let (program_name, status) = with_string_buffer(buffer_size, |buffer, max_size| {
+            self.low.EnumTrackMIDIProgramNamesEx(
+                project.to_raw(),
+                track.as_ptr(),
+                program_number,
+                buffer,
+                max_size,
+            )
+        });
+        match status {
+            true => Some(program_name),
+            false => None,
+        }
+    }
+
+    /// Gets note name for the note, if set on track.
+    pub fn get_track_midi_note_name<'a>(
+        &self,
+        track_index: u32,
+        pitch: u32,
+        channel: u32,
+    ) -> Option<&'a ReaperStr>
```

Comment:
Whenever you return a string that's owned by REAPER and can lose validity, use continuation-passing style. See `enum_project_markers_3` for example.